src/components/mutation/de.rs

```diff
@@ -0,0 +1,85 @@
+//! Mutation components for Differential Evolution (DE).
+
+use color_eyre::Section;
+use eyre::{ensure, eyre};
+use itertools::{multizip, Itertools};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    component::ExecResult, components::Component, population::AsSolutionsMut,
+    problems::VectorProblem, utils::with_index, State,
+};
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct DEMutation {
+    y: u32,
+    f: f64,
+}
+
+impl DEMutation {
+    pub fn from_params(y: u32, f: f64) -> ExecResult<Self> {
+        ensure!(
+            [1, 2].contains(&y),
+            "`y` needs to be one of {{1, 2}}, but was {}",
+            y
+        );
+        ensure!(
+            (0.0..=2.0).contains(&f),
+            "`f` must be in [0, 2], but was {}",
+            f
+        );
```

Comment:
> Would it be possible to have something like eyre contracts?

I think it is possible, I looked into the [`contracts`](https://docs.rs/contracts/latest/contracts/) crate a while ago and you need only add some additional cases to use `ensure!` instead of `assert!`. Would be really nice, in any case, but I don't have the time right now to look into this further.